picard/browser/browser.py

```diff
@@ -25,115 +25,169 @@
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
-
+from http.server import (
+    BaseHTTPRequestHandler,
+    HTTPServer,
+)
+import re
+import threading
 from urllib.parse import (
     parse_qs,
     urlparse,
 )
 
-from PyQt5 import QtNetwork
+from PyQt5 import QtCore
 
-from picard import log
+from picard import (
+    PICARD_APP_NAME,
+    PICARD_ORG_NAME,
+    PICARD_VERSION_STR,
+    log,
+)
 from picard.config import get_config
 from picard.util import mbid_validate
+from picard.util.thread import to_main
+
+
+SERVER_VERSION = '%s-%s/%s' % (PICARD_ORG_NAME, PICARD_APP_NAME, PICARD_VERSION_STR)
+RE_VALID_ORIGINS = re.compile(r'^(?:[^\.]+\.)*musicbrainz\.org$')
 
 
-def response(code):
-    if code == 200:
-        resp = '200 OK'
-    elif code == 400:
-        resp = '400 Bad Request'
-    else:
-        resp = '500 Internal Server Error'
-    return bytearray(
-        'HTTP/1.1 {}\r\n'
-        'Cache-Control: max-age=0\r\n'
-        '\r\n'
-        'Nothing to see here.\r\n'.format(resp), 'ascii')
+def _is_valid_origin(origin):
+    try:
+        url = urlparse(origin)
+    except ValueError:
+        return False
+    hostname = url.hostname
+    if not hostname:
+        return False
+    if RE_VALID_ORIGINS.match(hostname):
+        return True
+    config = get_config()
+    return config.setting['server_host'] == hostname
 
 
-class BrowserIntegration(QtNetwork.QTcpServer):
+class BrowserIntegration(QtCore.QObject):
 
-    """Simple HTTP server for web browser integration."""
+    listen_port_changed = QtCore.pyqtSignal(int)
 
     def __init__(self, parent=None):
         super().__init__(parent)
-        self.newConnection.connect(self._accept_connection)
-        self.port = 0
-        self.host_address = None
+        self.server = None
+
+    @property
+    def host_address(self):
+        if not self.server:
+            return ''
+        return self.server.server_address[0]
+
+    @property
+    def port(self):
+        if not self.server:
+            return 0
+        return self.server.server_address[1]
 
     def start(self):
-        if self.port:
+        if self.server:
             self.stop()
 
         config = get_config()
         if config.setting["browser_integration_localhost_only"]:
-            self.host_address = QtNetwork.QHostAddress(QtNetwork.QHostAddress.LocalHost)
+            host_address = '127.0.0.1'
         else:
-            self.host_address = QtNetwork.QHostAddress(QtNetwork.QHostAddress.Any)
+            host_address = '0.0.0.0'  # nosec
```

Comment:
It makes me think we could prolly add an option to specify listening address instead of listening on any. Something like if `browser_integration_localhost_only` is False, use `browser_integration_ip_address`, which defaults to `0.0.0.0`.
It would make sense for machines connected to multiple local networks.

Reply:
Yes, good idea. But for sure rather an exception for the majority of users (but so is the use case of enabling external networks at all)